### PR TITLE
Fix bug map key cannot be non string type

### PR DESCRIPTION
**Pull request by oliviazhang0809**

### CODE REVIEW

File: codegen/type_converter.go [hunk 1]
```diff
@@ -346,16 +346,22 @@ func (c *TypeConverter) genConverterForMap(
 		return err
 	}
 
+	valueStruct, isStruct := toFieldType.ValueSpec.(*compile.StructSpec)
+	sourceIdentifier := fromIdentifier
 	_, isStringKey := toFieldType.KeySpec.(*compile.StringSpec)
+
 	if !isStringKey {
-		return errors.Errorf(
```
Comment (Raynos) [inline codegen/type_converter.go]: We need a new check here. We do not want to generate code for `map<map<string,string>, string>` 

The key should be a `string` or a typedef for a `string`.

There is a `compile.RootTypeSpec()` method that will take a typedef and give you back the actual type it is.

File: codegen/type_converter_test.go [hunk 2]
```diff
@@ -1102,28 +1102,27 @@ func TestConvertWithBadImportMapOfBadStruct(t *testing.T) {
...
 	)
 
-	assert.Error(t, err)
-	assert.Equal(t, "", lines)
-	assert.Equal(t,
-		"could not convert key (one), map is not string-keyed.",
-		err.Error(),
-	)
+	assert.NoError(t, err)
+	assert.Equal(t, "out.One = make(map[structs.UUID]string, len(in.One))\nfor key, value := range in.One {\n\tout.One[ structs.UUID(key)] = string(value)\n}\nout.Two = make(map[int32]string, len(in.Two))\nfor key, value := range in.Two {\n\tout.Two[ int32(key)] = string(value)\n}", lines)
```
Comment (Raynos) [inline codegen/type_converter_test.go]: for reviewability and sanity please use 

    trim(`
        // code with whitespace
    `)

**Comment by ChuntaoLu** [comment]
For git commit hygiene, you may want to squash the commits of the same message or update the message if each commit does different thing.

screen shot 2017-10-12 at 3 09 59 pm

File: codegen/type_converter.go [hunk 3]
```diff
@@ -346,16 +346,30 @@ func (c *TypeConverter) genConverterForMap(
...
-			"could not convert key (%s), map is not string-keyed.",
-			toField.Name,
+		_, isMapKey := toFieldType.KeySpec.(*compile.MapSpec)
+		if isMapKey {
```
Comment (Raynos) [inline codegen/type_converter.go]: Sorry map key was an example. Doing `map<set<string>, string>` ; `map<list<string>, string` will also not work.

We probably need to check if the typedef resolves to `*compile.StringSpec` and reject every other type that's not string or typedef to string.

Comment (oliviazhang0809) [reply codegen/type_converter.go]: basically just collections won't work right? Do you have an example how we can leverage compile.RootTypeSpec()?

Comment (Raynos) [reply codegen/type_converter.go]: There should be examples of it. `git grep RootTypeSpec`

File: codegen/type_converter.go [hunk 4]
```diff
@@ -346,16 +346,34 @@ func (c *TypeConverter) genConverterForMap(
...
-		return errors.Errorf(
-			"could not convert key (%s), map is not string-keyed.",
-			toField.Name,
-		)
+		realType := compile.RootTypeSpec(toFieldType.KeySpec)
+		switch realType.(type) {
+		case
+			*compile.MapSpec,
+			*compile.SetSpec,
+			*compile.ListSpec:
```
Comment (Raynos) [inline codegen/type_converter.go]: We cannot support `*compile.StructSpec` either. 

I'm pretty sure `*compile.Int8Spec` , `*compile.DoubleSpec` and `*compile.BoolSpec` will also fail.

File: codegen/type_converter.go [hunk 5]
```diff
@@ -346,16 +346,38 @@ func (c *TypeConverter) genConverterForMap(
...
+
 	if !isStringKey {
-		return errors.Errorf(
-			"could not convert key (%s), map is not string-keyed.",
-			toField.Name,
-		)
+		realType := compile.RootTypeSpec(toFieldType.KeySpec)
+		switch realType.(type) {
+		case
+			*compile.MapSpec,
```
Comment (Raynos) [inline codegen/type_converter.go]: The full list is here ( https://github.com/uber/zanzibar/blob/master/codegen/thrift.go#L201-L230 )

For hygiene purposes and handling all cases it's better to panic in the `default` case and to explicitely handle the `*compile.StringSpec` case.

**Pull request merged**